packages/mdc-data-table/adapter.ts

```diff
@@ -0,0 +1,158 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/**
+ * Defines the shape of the adapter expected by the foundation.
+ * Implement this adapter for your framework of choice to delegate updates to
+ * the component in your framework of choice. See architecture documentation
+ * for more details.
+ * https://github.com/material-components/material-components-web/blob/master/docs/code/architecture.md
+ */
+
+import {MDCDataTableRowSelectionChangedEventDetail} from './types';
+
+export interface MDCDataTableAdapter {
+  /**
+   * Adds a class name to row element at given row index excluding header row.
+   *
+   * @param rowIndex Index of row element excluding header row.
+   * @param cssClasses CSS Class string to add.
+   */
+  addClassAtRowIndex(rowIndex: number, cssClasses: string): void;
+
+  /**
+   * @return Row count excluding header row.
+   */
+  getRowCount(): number;
+
+  /**
+   * @return Array of row elements excluding header row.
+   */
+  getRowElements(): Element[];
+
+  /**
+   * Returns row id of row element at given row index based on `data-row-id` attribute on row element `tr`.
+   *
+   * @param rowIndex Index of row element.
+   * @return Row id of row element, returns `null` in absence of `data-row-id` attribute on row element.
+   */
+  getRowIdAtIndex(rowIndex: number): string | null;
+
+  /**
+   * Returns index of row element that contains give child element.
+   *
+   * @param el Child element of row element.
+   * @return Index of row element.
+   */
+  getRowIndexByChildElement(el: Element): number;
+
+  /**
+   * @return Selected row count.
+   */
+  getSelectedRowCount(): number;
+
+  /**
+   * @param rowIndex Index of row element.
+   * @return True if row checkbox at given row index is checked.
+   */
+  isCheckboxAtRowIndexChecked(rowIndex: number): boolean;
+
+  /**
+   * @return True if header row checkbox is checked.
+   */
+  isHeaderRowCheckboxChecked(): boolean;
+
+  /**
+   * @return True if table rows are selectable.
+   */
+  isRowsSelectable(): boolean;
+
+  /**
+   * Notifies when row selection is changed.
+   *
+   * @param data Event detail data for row selection changed event.
+   */
+  notifyRowSelectionChanged(data: MDCDataTableRowSelectionChangedEventDetail): void;
+
+  /**
+   * Notifies when header row is checked.
+   */
+  notifySelectedAll(): void;
+
+  /**
+   * Notifies when header row is unchecked.
+   */
+  notifyUnselectedAll(): void;
+
+  /**
+   * Initializes header row checkbox. Destroys previous header row checkbox instance if any.
+   * @return Can return Promise only if registering checkbox is asynchronous.
+   */
+  registerHeaderRowCheckbox(): Promise<void> | void;
```

Comment:
Does this actually return a promise? In the component the method is:

```
        if (this.headerRowCheckbox_) {
          this.headerRowCheckbox_.destroy();
        }

        const checkboxEl = (this.root_.querySelector(strings.HEADER_ROW_CHECKBOX_SELECTOR) as HTMLElement);
        this.headerRowCheckbox_ = this.checkboxFactory_(checkboxEl);
```

Reply:
It can optionally return `Promise`. See `layoutAsync` in foundation.ts